amacc.c

```diff
@@ -568,112 +546,135 @@ void expr(int lev)
+            next(); expr(Shl);
+            if (*n == Num && *b == Num) n[1] = b[1] >= n[1]; else { *--n = (int) b; *--n = Ge; }
+            ty = INT; 
+            break;
+        case Shl: 
+            next(); expr(Add);
+            if (*n == Num && *b == Num) {
+                if (n[1] < 0) n[1] = b[1] >> -n[1];
+                else n[1] = b[1] << n[1];
+            } else { *--n = (int) b; *--n = Shl; }
+            ty = INT; 
+            break;
+        case Shr: 
+            next(); expr(Add);
+            if (*n == Num && *b == Num){
+                if (n[1] < 0) n[1] = b[1] << -n[1];
+                else n[1] = b[1] >> n[1];
+            } else { *--n = (int) b; *--n = Shr; }
+            ty = INT; 
             break;
-        case Gt:  next(); *++e = PSH; expr(Shl); *++e = GT;  ty = INT; break;
-        case Le:  next(); *++e = PSH; expr(Shl); *++e = LE;  ty = INT; break;
-        case Ge:  next(); *++e = PSH; expr(Shl); *++e = GE;  ty = INT; break;
-        case Shl: next(); *++e = PSH; if (tk == Sub) { tk = Add; expr(Add); *++e = SHR; }
-                                      else { expr(Add); *++e = SHL; }
-                                      ty = INT; break;
-        case Shr: next(); *++e = PSH; if (tk == Sub) { tk = Add; expr(Add); *++e = SHL; }
-                                      else { expr(Add); *++e = SHR; }
-                                      ty = INT; break;
         case Add:
-            next(); *++e = PSH; expr(Mul);
-            sz = (ty = t) >= PTR2 ? sizeof(int) :
-                                    ty >= PTR ? tsize[ty - PTR] : 1;
-            if (sz > 1) { *++e = PSH; *++e = IMM; *++e = sz; *++e = MUL; }
-            *++e = ADD;
+            next(); expr(Mul);
+            sz = (ty = t) >= PTR2 ? sizeof(int) :  ty >= PTR ? tsize[ty - PTR] : 1;
+            if (*n == Num) n[1] *= sz; 
+            if (*n == Num && *b == Num) n[1] += b[1]; else { *--n = (int) b; *--n = Add; }
             break;
         case Sub:
-            next(); *++e = PSH; expr(Mul);
-            sz = t >= PTR2 ? sizeof(int) :
-                             t >= PTR ? tsize[t - PTR] : 1;
-            if (t == ty && sz > 1) {
-                switch (sz) {
-                case 4: *++e = SUB; *++e = PSH; *++e = IMM; *++e = 2; *++e = SHR; break;
-                default: *++e = SUB; *++e = PSH; *++e = IMM; *++e = sz; *++e = SUB; break;
+            next(); expr(Mul);
+            sz = t >= PTR2 ? sizeof(int) : t >= PTR ? tsize[t - PTR] : 1;
+            if (sz > 1 && *n == Num){ *--n = sz; *--n = Num; --n; *n = (int) (n + 3); *--n = Mul; }
+            if (*n == Num && *b == Num) n[1] -= b[1]; 
+            else { 
+                *--n = (int) b; *--n = Sub; 
+                if (t == ty && sz > 1) {
+                    switch (sz) {
+                    case 4: *--n = 2; *--n = Num; --n; *n = (int) (n + 3);  *--n = Shr; break;
+                    default: *--n = sz; *--n = Num; --n; *n = (int) (n + 3); *--n = Sub; break;
```

Comment:
For `default` case, you don't have to append `break` by the end of these statements.